sqstaskmaster/local.py

```diff
@@ -0,0 +1,138 @@
+        logger.info("noop delete")
+
+    def __str__(self):
+        return "LocalMessage(body: {}; attributes: {}; message_attributes: {})".format(
+            self.body, self.attributes, self.message_attributes
+        )
+
+
+class LocalQueue:
+    """
+    Partial implementation of the SQS Queue interface for local integration testing of the producer and consumer.
+    This does not implement behavior beyond the ability to put and get messages in FIFO order to a queue named by a url.
+
+    Some of the SQS features that are missing: Retries, visibility & timeout, thread/process safety.
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html#queue
+    """
+
+    local_queues = defaultdict(queue.SimpleQueue)
+
+    def __init__(self, url):
+        self.url = url
+        self._queue = self.local_queues[url]
+
+    def load(self):
+        logger.info("Noop load!")
+
+    @property
+    def queue(self):
+        return self._queue
+
+    @property
+    def attributes(self):
+        logger.info("Noop attributes!")
+        # Return partial set of attributes
+        return {
+            "QueueArn": "LocalQueue replacing: " + self.url,
+            "ApproximateNumberOfMessages": self.queue.qsize(),
+            "ApproximateNumberOfMessagesNotVisible": 0,
+            "ApproximateNumberOfMessagesDelayed": 0,
+            "FifoQueue": True,
+            "ContentBasedDeduplication": False,
+        }
+
+    def purge(self):
+        while True:
+            try:
+                self.queue.get_nowait()
+            except queue.Empty:
+                break
+
+    def send_message(self, **kwargs):
+        """
+        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html#SQS.Queue.send_message
+        :param kwargs: expects MessageBody (String) and optional MessageAttributes (Dict)
+        :return: partial metadata of actual API
+        """
+        attributes = {}
+        # Ignore attribute behavior which is not relevant
+        message = LocalMessage(
+            kwargs["MessageBody"], attributes, kwargs.get("MessageAttributes", {})
+        )
+
+        logger.debug("Sending message: %s", message)
+        self.queue.put_nowait(message)
+        return {"MD5OfMessageBody": "Fake LocalMessage MD5 Body"}
+
+    def receive_messages(self, **kwargs):
+        """
+        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html#SQS.Queue.receive_messages
+        :param kwargs: simulates behavior of WaitTimeSeconds and MaxNumberOfMessages
+        :return:
+        """
+        time_remaining = kwargs.get("WaitTimeSeconds", 20)
+
+        result = []
+
+        tic = time.perf_counter()
+        for i in range(kwargs.get("MaxNumberOfMessages", 1)):
+            time_remaining = max(0, time_remaining - time.perf_counter() + tic)
+
+            try:
+                result.append(self.queue.get(block=True, timeout=time_remaining))
```

Comment:
and this is different from `purge` in that (a) waiting for a response is allowed for a finite time, (b) the response is used (it is appended to `result`), and (c) this call is made only once, rather than in a `while True` loop

Reply:
yes - get upto `kwargs.get("MaxNumberOfMessages", 1)` messages
Wait upto kwargs.get("WaitTimeSeconds", 20) to get them